### ISSUE TITLE
Diffraction Screen: Code Review

### ISSUE THREAD

**Issue by samreid** [phetsims/wave-interference#376]
@ariel-phet said:

> When do you think diffraction screen will be in a place ready for code review?  We have some minor polishes to do (discuss tomorrow), but I think after that no changes are anticipated

I replied:

> If there are no other changes, I would just need 1-2 hours to double check and make sure everything’s all set.

@ariel-phet said:

> Cool, CM will be the one doing the review

I enumerated a number of steps in #372 that may need to be accomplished before I turn this over to @pixelzoom for code review.  Putting on hold until then.

**Comment by samreid** [comment]
@ariel-phet said late next week would work well.

**Comment by samreid** [comment]
#375 and #377 and #383 will need to be addressed before we can go to code review, so this will be on hold longer.

**Comment by samreid** [comment]
Instructions to the reviewer:

- [ ] The new screen to be reviewed is wave-interference/js/diffraction, approximately 1500 lines of code
- [ ] Complete review of https://github.com/phetsims/scenery-phet/issues/506
- [ ] Please visit `TODO: to the reviewer` comments in this directory and address or comment on them

* Note there is ongoing work as part of the "Waves Intro" sonification project elsewhere in this repo, you may see TODOs or other in-progress work related to that (outside of the diffraction directory)
* Note there are still some outstanding issues in this repo marked for label `wave-interference:2.0`, but these are not expected to hamper the code review, so this is ready to go.

**Comment by pixelzoom** [comment]
@samreid What's the priority? I may not get to this until > July 2.  If that's a problem, consult with @ariel-phet about an alternate reviewer.

**Comment by samreid** [comment]
@ariel-phet replied over slack

**Comment by pixelzoom** [comment]
On Slack, @ariel-phet said:
> Reasonably high, if you can work on it week of 6/24 great. But still lower priority than the iO sprint

**Comment by pixelzoom** [comment]
Review is completed.   Looking good overall.  And this was a fun screen to play around with.

* I added a few missing annotations in the above commits.

* I added a few `REVIEW` comments inline.

* It's typically up to developer discretion to put some view-specific things in the model when convenient. But I didn't expect to find things like `CanvasRenderingContext2D` in the model. And the primary responsibility of the model appears to be rendering  the view using Canvas.  This makes it difficult to see which bits of the model are independent of view (I'm guess just the Properties?).    This is probably not something that you'd want to change at this late point, so consider documenting (perhaps in implementation-notes.md?) why the model contains so much stuff that's specific to Canvas. 

* Re the `TODO: to the reviewer` in `DiffractionScene`... If I understand correctly, you want to update the canvases when any Property associated with a scene changes.  If that's correct, then consider these changes:
(1) Pass the value of `this.properties` into the `DiffractionScene ` constructor, instead of having subclasses set it after the fact, which feels a little unsafe.  I.e.:

```js
 class DiffractionScene {

    // @param {Property.<*>[]} properties
    constructor( properties ) {
      this.properties = properties; // @private
...
```

Example subclass:

```js
  class CircleSquareScene extends DiffractionScene {

    constructor() {
      
      // in mm
      const circleDiameterProperty = new NumberProperty( 50 * 1E-3, {
        range: new Range( 40 * 1E-3, 150 * 1E-3 )
      } );

      // in mm
      const squareWidthProperty = new NumberProperty( 50 * 1E-3, {
        range: new Range( 40 * 1E-3, 150 * 1E-3 )
      } );
      
      super( [ circleDiameterProperty, squareWidthProperty ] );

      // @public
      this.circleDiameterProperty = circleDiameterProperty;
      this.squareWidthProperty = squareWidthProperty;
    }
```

(2) Add an Emitter to `DiffractionScene` constructor, to notify listeners when a scene has changed:

```
// @public
this.changedEmitter = new Emitter();
```

If you want to make it private, add method `addChangedListener`:

```js
addChangedListener( listener ) {
  this.changedEmitter.addListener( listener );
}
```

(3) Make `DiffractionScene` connect a listener to each Property in `this.properties` that calls `this.changedEmitter.emit`.  This makes `DiffractionScene` responsible for the specifics of what "changed" means. In `DiffractionScene` constructor:

```js
const changedListener = () => { this.changedEmitter.emit(); }
this.properties.forEach( property => property.link( changedListener ) );
```
(4) In `DiffractionScreenView`, replace 
`model.scenes.forEach( scene => scene.link( updateCanvases ) );` 
with 
`model.scenes.forEach( scene => scene.changedEmitter.addListener( updateCanvases ) );`
Or if you wanted to make `changedEmitter` private, you could have:
`model.scenes.forEach( scene => scene.addChangeListener( updateCanvases ) );`

**Comment by pixelzoom** [comment]
In the previous comment, I said:
> ... consider these changes:

Thinking about this more, steps 2-4 above are really unnecessary.  I guess what I don't like is the method `link` in DiffractionScene` --  it makes it look like a `Property` when it's not.   Rename it to something like `addChangedListener` and that would address my concern.  

I think step 1 still a good idea, since making a subclass responsible for superclass field assignment seems like something that's good to avoid in general.

**Comment by samreid** [comment]
> But I didn't expect to find things like CanvasRenderingContext2D in the model.

This simulation uses a canvas in the model for generating the model aperture shape.  It is described in DiffractionScene.js like so:

> // Render to a canvas and sample points.  Using kite Shape.containsPoint on the SVG shape declaration was much too slow

and in the implementation-notes.md like so:

> There are 5 apertures (each shown in a scene) with different tunable characteristics.  The aperture renders to a canvas in the model for the sake of determining which points in the aperture are open and closed.  The same "render to canvas"  code is used for rendering crisp images in the view.

Please let me know what I can do to make this clearer.

**Comment by samreid** [comment]
> I think step 1 is still a good idea, since making a subclass responsible for superclass field assignment seems like something that's good to avoid in general.

Addressed in preceding commits.

**Comment by samreid** [comment]
> I guess what I don't like is the method `link` in `DiffractionScene` --  it makes it look like a `Property` when it's not.   Rename it to something like `addChangedListener` and that would address my concern.  

In the preceding commit, I renamed it to `linkToAllProperties`, will that work?

**Comment by samreid** [comment]
I've addressed the `REVIEW` comments in the code and responded to all comments in this issue.  Back to @pixelzoom for discussion.

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/wave-interference/commit/e1aee1a573545f45f22ec278c8b3ee07d12eeedb, `this.wavelengthProperty` in `DiffractionModel` was changed from 'nm' to 'mm'.   It's default value is `WaveInterferenceConstants.DEFAULT_WAVELENGTH`, which is documented as "in nm".  So something seems incorrect here.

**Comment by samreid** [comment]
Thanks!  I fixed it back to nm.

**Comment by pixelzoom** [comment]
> In the preceding commit, I renamed it to `linkToAllProperties`, will that work?

Works for me.

**Comment by pixelzoom** [comment]
Re https://github.com/phetsims/wave-interference/issues/376#issuecomment-505559241...  Thanks, I did see those bits of documentation.  They tell me (as you said) that a canvas is used in the model for generating the model aperture shape.  But they don't tell me _why_ that's appropriate to do in the model, or why a shape belongs in the model.  A shape is (typically) one view-specific way of representing something, but maybe that's not the case here.  For the benefit of those of us who don't have a lot of experience in this domain, it might be useful to say something about this.

**Comment by samreid** [comment]
I elaborated on the model canvas in the preceding commit.  @pixelzoom can you please review?

**Comment by pixelzoom** [comment]
Thanks, that's helpful.  

This issue seems to be done, so closing.  Feel free to reopen if I missed something.
